modules/javafx.graphics/src/main/java/javafx/css/SimpleSelector.java

```diff
@@ -323,8 +378,8 @@ private boolean matchStyleClasses(StyleClassSet otherStyleClasses) {
     @Override public int hashCode() {
         int hash = 7;
         hash = 31 * (hash + name.hashCode());
-        hash = 31 * (hash + styleClassSet.hashCode());
-        hash = 31 * (hash + styleClassSet.hashCode());
+        hash = 31 * (hash + selectorStyleClassNames.hashCode());
+        hash = 31 * (hash + selectorStyleClassNames.hashCode());
```

Comment:
this pattern destroys some information that otherwise should be used to produce a better hash code.
instead of 
`hash = 31 * (hash + newPart);`
it should be
`hash = 31 * hash + newPart;`

(also in CompoundSelector)

Reply:
I agree that it is not nice :)  However, changing this seems out of scope for this PR as my changes are unrelated to this, and I didn't even touch `CompoundSelector`.  As selectors are never stored in sets/maps, it is hard to justify changing this as part of this PR as it would not be related to improving performance.

Reply:
these lines discard about 4.5 bits of information on each step, surely that will affect performance, no?

Reply:
I am confused, and maybe I am missing something. As far as I know, this method is not called anywhere. I put a breakpoint in it. JavaFX does not use this method anywhere, nor are Selectors ever used as keys in Sets or Maps.

What am I missing?

Reply:
you are right, this code does not affect performance (I could not hit a break point here either).
still, since you are touching these lines, why not do it right?

Reply:
I can undo the rename so they're not touched.  My problem with fixing this is that I then also should do it for `CompoundSelector`, which is just completely unrelated to the intent of this PR.  I'm pretty sure that within JavaFX we don't change lines that are unrelated to the intent of the change (these lines were only hit because of a clarifying field rename related to this change).

Normally I would do such fixes in an extra commit, where the first commit would be something like "fix: Improve selector match performance" and the second is "fix: Fix bug in Selectors hashCode algorithm"...

... but since everything is squashed as the way of working in FX (instead of having stand alone commits that are fast forwarded when included in master), it would have to be done the official way, which is file a ticket, create a new PR, have it also reviewed, etc.  This puts an additional burden on the already long review queue, with well intentioned solid PR's that are abandoned because of lack of feedback and reviews... Even simple things like fixing warnings I've given up on as I see it taking time away from more pressing issues.

Reply:
I feel the same kind of frustration sometimes.  The review process is indeed painfully slow.

A different JBS/PR would be fine, I think, something along the lines of "suboptimal hashCode() implementation in ..."

Reply:
I agree with John that changing the existing hash algorithm it really out of scope for this PR (even if it is suboptimal, which it is).

Filing a separate bug he advantage of decoupling it in time as well as not burdening this review. It may or may not even be worth implementing the optimization if no one uses it, but in any case, it can be done later.